### PR TITLE
redo(ticdc): owner and processores shouldn't share one redo log writer

### LINKED ISSUE #6695: redo log: change stuck in case redo_apply_multiple_workload_node_restart_sync
### What did you do?

- create changefeed1: enable redo log, filter sysbench table
- create changefeed2: enable redo log, filter go-ycsb table
- run sysbench prepare, meanwhile, all cdc node restart
- run truncate all sysbench table
- run sysbench run, meanwhile, all pd restart
- drop all sysbench tables
- run sysbench prepare again
- run go-ycsb workload
- sendfinishmark and waitforthisDDL

### What did you expect to see?

_No response_

### What did you see instead?

k0JvrUqnLJ

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
(paste TiDB cluster version here)
```

Upstream TiKV version (execute `tikv-server --version`):

```console
(paste TiKV version here)
```

TiCDC version (execute `cdc version`):

```console
(paste TiCDC version here)
Release Version: v6.1.0-20220809
Git Commit Hash: 0c82c3405e49ac1da9f5e705d3862ec5a79c22bf
Git Branch: heads/refs/tags/v6.1.0-20220809
UTC Build Time: 2022-08-09 13:29:30
Go Version: go version go1.18.5 linux/amd64
Failpoint Build: false
```

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What problem does this PR solve?

Issue Number: close #6695 

### What is changed and how it works?

Reusing log write in redo manager should be in a correct way

#### Tests 

 - Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/redo/manager.go [hunk 1]
```diff
@@ -533,20 +533,35 @@ func (m *ManagerImpl) onResolvedTsMsg(tableID model.TableID, resolvedTs model.Ts
...
+			zap.String("namespace", m.changeFeedID.Namespace),
+			zap.String("changefeed", m.changeFeedID.ID),
+			zap.Error(err))
 		select {
 		case logErrCh <- err:
 		default:
 		}
 	}
 
+	log.Info("redo manager bgUpdateLog is running",
```
Comment (nongfushanquan) [inline cdc/redo/manager.go:546]: too many logs will appear in the log files

**Review by nongfushanquan — COMMENTED** [review]
please and an issue

File: cdc/redo/manager.go (hunk 1 above)
Comment (hicqu) [reply cdc/redo/manager.go:570]: There will be one this log per changefeed. So I guess it's acceptable.

File: cdc/redo/manager.go [hunk 2]
```diff
@@ -349,6 +364,16 @@ func (m *ManagerImpl) UpdateResolvedTs(
 	return nil
 }
 
+// UpdateCheckpointTs updates checkpoint to the manager.
+func (m *ManagerImpl) UpdateCheckpointTs(ckpt model.Ts) {
+	for {
+		curr := atomic.LoadUint64(&m.checkpointTs)
+		if ckpt <= curr || atomic.CompareAndSwapUint64(&m.checkpointTs, curr, ckpt) {
```
Comment (liuzix) [inline cdc/redo/manager.go:371]: When will `ckpt <= curr` happen?

File: cdc/redo/manager.go [hunk 3]
```diff
@@ -567,16 +600,55 @@ func (m *ManagerImpl) bgUpdateLog(ctx context.Context, errCh chan<- error) {
...
+			break
+		}
+	}
+
+	// NOTE: the goroutine should never exit until the err is put into errCh successfully
+	// or the context is canceled.
+	select {
+	case <-ctx.Done():
+	case errCh <- err:
+	}
```
Comment (liuzix) [inline cdc/redo/manager.go:625]: I think this is a deadlock risk.

File: cdc/owner/changefeed.go [hunk 4]
```diff
@@ -479,12 +483,15 @@ LOOP:
...
+	redoManagerOpts := &redo.ManagerOptions{EnableBgRunner: true, ErrCh: c.errCh}
+	mgr, err := redo.NewManager(stdCtx, c.state.Info.Config.Consistent, redoManagerOpts)
+	c.redoManager = mgr
 	if err != nil {
 		return err
 	}
-	c.redoManager = redoManager
+	log.Info("owner creates redo manager",
+		zap.String("namespace", contextutil.ChangefeedIDFromCtx(stdCtx).Namespace),
+		zap.String("namespace", contextutil.ChangefeedIDFromCtx(stdCtx).ID))
```
Comment (3AceShowHand) [inline cdc/owner/changefeed.go:494]:
```suggestion
zap.String("namespace", c.id.Namespace),
				zap.String("changefeed", c.id.ID))
```

File: cdc/processor/processor.go [hunk 5]
```diff
@@ -703,6 +703,9 @@ func (p *processor) lazyInitImpl(ctx cdcContext.Context) error {
 	if err != nil {
 		return err
 	}
+	log.Info("processor creates redo manager",
+		zap.String("namespace", contextutil.ChangefeedIDFromCtx(stdCtx).Namespace),
```
Comment (3AceShowHand) [inline cdc/processor/processor.go:707]:
```suggestion
		zap.String("namespace", p.changefeedID.Namespace,
		zap.String("changefeed", p.changefeedID.ID)
```

File: cdc/redo/manager.go [hunk 6]
```diff
@@ -532,27 +566,26 @@ func (m *ManagerImpl) onResolvedTsMsg(tableID model.TableID, resolvedTs model.Ts
 
 func (m *ManagerImpl) bgUpdateLog(ctx context.Context, errCh chan<- error) {
 	logErrCh := make(chan error, 1)
-	handleErr := func(err error) {
-		select {
-		case logErrCh <- err:
-		default:
-		}
-	}
+	handleErr := func(err error) { logErrCh <- err }
```
Comment (3AceShowHand) [inline cdc/redo/manager.go:569]: please add a comment about your consideration here, why a buffered channel for error ?

File: cdc/redo/manager.go [hunk 7]
```diff
@@ -254,18 +264,26 @@ func NewManager(ctx context.Context, cfg *config.ConsistentConfig, opts *Manager
...
 
-	if opts.EnableBgRunner {
+	// TODO: better to wait background goroutines after the context is canceled.
+	if m.opts.EnableBgRunner {
 		go m.bgUpdateLog(ctx, opts.ErrCh)
 	}
+	if m.opts.EnableGCRunner {
+		go m.bgGC(ctx)
+	}
+
```
[the remark this answers is not in the document]
Comment (hicqu) [reply cdc/redo/manager.go:286]: There is no explicit API to close `ManagerImpl`. Now background goroutines are controled with `context.Context`.

File: cdc/redo/manager.go (hunk 2 above)
Comment (hicqu) [reply cdc/redo/manager.go:371]: It happends when the function is called in multiple threads.

File: cdc/redo/manager.go (hunk 3 above)
Comment (hicqu) [reply cdc/redo/manager.go:625]: Owner and processor will cancel `ctx` as expected so here the `select` can finish finally.

File: cdc/processor/pipeline/sink.go [hunk 8]
```diff
@@ -158,16 +158,14 @@ func (n *sinkNode) flushSink(ctx context.Context, resolved model.ResolvedTs) (er
 		resolved = model.NewResolvedTs(currentBarrierTs)
 	}
 	if n.redoManager != nil && n.redoManager.Enabled() {
-		redoTs := n.redoManager.GetMinResolvedTs()
 		redoFlushed := n.redoManager.GetResolvedTs(n.tableID)
-		if currentBarrierTs > redoTs {
+		if currentBarrierTs > redoFlushed {
```
Comment (nongfushanquan) [inline cdc/processor/pipeline/sink.go:162]: why not merge this if condition into the next one

File: cdc/processor/pipeline/sink.go [hunk 9]
```diff
@@ -211,6 +209,10 @@ func (n *sinkNode) flushSink(ctx context.Context, resolved model.ResolvedTs) (er
 	}
 	n.checkpointTs.Store(checkpoint)
 
+	if n.redoManager != nil && n.redoManager.Enabled() {
```
Comment (nongfushanquan) [inline cdc/processor/pipeline/sink.go:212]: //todo : fix me, replace the checkpoint with the global checkpoint

File: cdc/redo/manager.go [hunk 10]
```diff
@@ -532,27 +566,26 @@ func (m *ManagerImpl) onResolvedTsMsg(tableID model.TableID, resolvedTs model.Ts
...
 		case <-ctx.Done():
 			return
-		case err := <-logErrCh:
-			select {
-			case errCh <- err:
-			default:
-				log.Error("err channel in redoManager is full", zap.Error(err))
-			}
-			return
+		case err = <-logErrCh:
```
Comment (nongfushanquan) [inline cdc/redo/manager.go:590]: Pls add log about the err

File: cdc/owner/changefeed.go [hunk 11]
```diff
@@ -479,12 +483,15 @@ LOOP:
...
+	redoManagerOpts := &redo.ManagerOptions{EnableBgRunner: true, ErrCh: c.errCh}
+	mgr, err := redo.NewManager(stdCtx, c.state.Info.Config.Consistent, redoManagerOpts)
+	c.redoManager = mgr
 	if err != nil {
 		return err
 	}
-	c.redoManager = redoManager
+	log.Info("owner creates redo manager",
+		zap.String("namespace", c.id.Namespace),
+		zap.String("namespace", c.id.ID))
```
Comment (3AceShowHand) [inline cdc/owner/changefeed.go:494]:
```suggestion
		zap.String("changefeed", c.id.ID))
```

File: cdc/processor/processor.go [hunk 12]
```diff
@@ -703,6 +708,9 @@ func (p *processor) lazyInitImpl(ctx cdcContext.Context) error {
 	if err != nil {
 		return err
 	}
+	log.Info("processor creates redo manager",
+		zap.String("namespace", p.changefeedID.Namespace),
+		zap.String("namespace", p.changefeedID.ID))
```
Comment (3AceShowHand) [inline cdc/processor/processor.go:713]:
```suggestion
		zap.String("changefeed", p.changefeedID.ID))
```

File: cdc/redo/manager.go (hunk 2 above)
Comment (nongfushanquan) [reply cdc/redo/manager.go:371]: if global checkpoint is set here,, this condition will not happen?

File: cdc/redo/manager.go [hunk 13]
```diff
@@ -567,16 +602,66 @@ func (m *ManagerImpl) bgUpdateLog(ctx context.Context, errCh chan<- error) {
...
+	}
+}
+
+func (m *ManagerImpl) bgGC(ctx context.Context) {
+	ticker := time.NewTicker(time.Duration(defaultGCIntervalInMs) * time.Millisecond)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("redo manager GC exits as context canceled",
```
Comment (zhaoxinyu) [inline cdc/redo/manager.go:646]:
```suggestion
			log.Info("redo manager GC exits as context cancelled",
```

File: cdc/redo/options.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+	return &ManagerOptions{
+		EnableBgRunner: true,
+		EnableGCRunner: false,
+		EmitMeta:       true,
+		EmitRowEvents:  false,
+		EmitDDLEvents:  true,
+	}
+}
+
+// NewProcessorManagerOptions creates a maanger options for processor.
```
Comment (zhaoxinyu) [inline cdc/redo/options.go:47]:
```suggestion
// NewProcessorManagerOptions creates a manager options for processor.
```

File: cdc/redo/options.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+	return &ManagerOptions{
+		EnableBgRunner: true,
+		EnableGCRunner: true,
+		EmitMeta:       false,
+		EmitRowEvents:  true,
+		EmitDDLEvents:  false,
+	}
+}
+
+// newMockManagerOptions creates a maanger options for mock tests.
```
Comment (zhaoxinyu) [inline cdc/redo/options.go:58]:
```suggestion
// newMockManagerOptions creates a manager options for mock tests.
```

File: cdc/redo/writer/writer.go [hunk 16]
```diff
@@ -476,29 +423,42 @@ var getAllFilesInS3 = func(ctx context.Context, l *LogWriter) ([]string, error)
...
+	}
+	if l.cfg.EmitMeta {
+		err = multierr.Append(err, l.flushLogMeta(checkpointTs, resolvedTs))
+	}
+	return
 }
 
 func (l *LogWriter) isStopped() bool {
-	return !l.ddlWriter.IsRunning() || !l.rowWriter.IsRunning()
+	var rowStoped, ddlStoped bool
```
Comment (zhaoxinyu) [inline cdc/redo/writer/writer.go:454]:
```suggestion
	var rowStopped, ddlStopped bool
```

File: cdc/redo/writer/writer.go [hunk 17]
```diff
@@ -476,29 +423,42 @@ var getAllFilesInS3 = func(ctx context.Context, l *LogWriter) ([]string, error)
...
+	if l.cfg.EmitRowEvents {
+		rowStoped = !l.rowWriter.IsRunning()
+	}
+	if l.cfg.EmitDDLEvents {
+		ddlStoped = !l.ddlWriter.IsRunning()
```
Comment (zhaoxinyu) [inline cdc/redo/writer/writer.go:459]:
```suggestion
		ddlStopped = !l.ddlWriter.IsRunning()
```

File: cdc/redo/writer/writer.go [hunk 18]
```diff
@@ -476,29 +423,42 @@ var getAllFilesInS3 = func(ctx context.Context, l *LogWriter) ([]string, error)
...
+	}
+	return rowStoped || ddlStoped
```
Comment (zhaoxinyu) [inline cdc/redo/writer/writer.go:461]:
```suggestion
	return rowStopped || ddlStopped
```

**Review by zhaoxinyu — APPROVED** [review]

**Review by 3AceShowHand — APPROVED** [review]

**Pull request merged**